mne/preprocessing/ica.py

```diff
@@ -2079,3 +2079,233 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+        0.6 and 0.95. In the original proposal, lower values are considered,
+        but this is not yet implemented.
+        If list of floats: search for the best map in the specified range of
+        correlation strengths.
+        If float > 0: select ICs correlating better than this.
+        If float > 1: use find_outliers to identify ICs within subjects (not in
+        original Corrmap)
+        Defaults to "auto".
+    name : str
+        Categorised ICs are stored in a default dictionary "labels". This
+        parameter gives the key under which found ICs will be stored.
+        Defaults to "bads".
+    plot : bool
+        Should constructed template and selected maps be plotted?
+    ch_type : 'mag' | 'grad' | 'planar1' | 'planar2' | 'eeg'
+            The channel type to plot. Defaults to 'eeg'.
+
+    Returns
+    -------
+    new_icas  :  list
+        Returns a list of fitted ICs, enrichened with the indices of the
+        selected maps in the field specified by the name keyword.
+    """
+
+    def vcorrcoef(X, y):  # vectorised correlation, by Jon Herman
+        Xm = np.reshape(np.mean(X, axis=1), (X.shape[0], 1))
+        ym = np.mean(y)
+        r_num = np.sum((X - Xm) * (y - ym), axis=1)
+        r_den = np.sqrt(np.sum((X - Xm) ** 2, axis=1) * np.sum((y - ym) ** 2))
+        r = r_num / r_den
+        return r
+
+    def get_ica_map(ica, components=None):
+        if components is None:
+            components = range(ica.n_components_)
+        maps = np.dot(ica.mixing_matrix_[:, components].T,
+                      ica.pca_components_[:ica.n_components_])
+        return maps
+
+    def find_max_corrs(all_maps, target, threshold):
+        all_corrs = [vcorrcoef(subj, target) for subj in all_maps]
+        abs_corrs = [np.abs(a) for a in all_corrs]
+        corr_polarities = [np.sign(a) for a in all_corrs]
+
+        if threshold <= 1:
+            max_corrs = [list(np.nonzero(s_corr > threshold)[0])
+                         for s_corr in abs_corrs]
+        else:
+            max_corrs = [list(find_outliers(s_corr, threshold=threshold))
+                         for s_corr in abs_corrs]
+
+        am = [l[i] for l, i_s in zip(abs_corrs, max_corrs)
+              for i in i_s]
+        median_corr_with_target = np.median(am)
+
+        polarities = [l[i] for l, i_s in zip(corr_polarities, max_corrs)
+                      for i in i_s]
+
+        maxmaps = [l[i] for l, i_s in zip(all_maps, max_corrs)
+                   for i in i_s]
+
+        try:
+            newtarget = np.zeros(maxmaps[0].size)
+            for maxmap, polarity in zip(maxmaps, polarities):
+                newtarget += maxmap * polarity
+
+            newtarget /= len(maxmaps)
+
+            sim_i_o = np.abs(np.corrcoef(target, newtarget)[1, 0])
+
+            return newtarget, median_corr_with_target, sim_i_o, max_corrs
+        except IndexError:
+            return [], 0, 0, []
```

Comment:
In my experience returning empty stuff often not a good idea. Can lead to
unnoticed bugs that propagate. I personally prefer clean errors, and I
think we don't have functions that return empty containers on failure.

2015-02-22 21:01 GMT+01:00 jona-sassenhagen <notifications@github.com>:

> In mne/preprocessing/ica.py
> <https://github.com/mne-tools/mne-python/pull/1795#discussion_r25135817>:
>
> > +
> > +        maxmaps = [l[i] for l, i_s in zip(all_maps, max_corrs)
> > +                   for i in i_s]
> > +
> > +        try:
> > +            newtarget = np.zeros(maxmaps[0].size)
> > +            for maxmap, polarity in zip(maxmaps, polarities):
> > +                newtarget += maxmap * polarity
> > +
> > +            newtarget /= len(maxmaps)
> > +
> > +            sim_i_o = np.abs(np.corrcoef(target, newtarget)[1, 0])
> > +
> > +            return newtarget, median_corr_with_target, sim_i_o, max_corrs
> > +        except IndexError:
> > +            return [], 0, 0, []
>
> Oh, this isn't an error, it's for dealing with the few, but expected
> iterations where no map is found. In most cases, these lines will work as
> in the try clause, but in a few situations, it is being fed with an empty
> map, and in that case it should return a 0. I was assuming since only the
> try part is run in most cases, it'd be better to ask for forgiveness for
> the exceptions than check every time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1795/files#r25135817>.
>